### ISSUE TITLE
[Testing] Discussion on the Kura testing framework

### ISSUE THREAD

**Issue by cdealti** [eclipse/kura#833]
Following the discussion in #800, here's is a first difference:

### UT in src/test/java of the org.eclipse.kura.camel plugin
* Tests run fine in CI with **maven-surefire-plugin** (OSGi is not needed)
* Tests cannot be run from Eclipse due to resolution errors for imported packages. Note that these packages are not imported in the MANIFEST of ``org.eclipse.kura.camel``.

![image](https://cloud.githubusercontent.com/assets/1225835/20604879/f5d01cd8-b268-11e6-91d2-0f48e91fa1cb.png)

### UT for the org.eclipse.kura.core.net plugin in src/test/java of the org.eclipse.kura.core.net.test test plugin
* Tests run fine in CI with **maven-surefire-plugin** (OSGi is not needed)
* Tests can be run from Eclipse. Dependency from JUnit, Mockito etc are only in the MANIFEST of the test plugin.
* The test plugin is a fragment of the plugin under test and can access its internal packages
![image](https://cloud.githubusercontent.com/assets/1225835/20604926/3200040c-b269-11e6-9573-e8218143e9d2.png)

**Comment by cdealti** [comment]
@ctron @djuro @markoer comments?

**Comment by markoer** [comment]
Another option is to add test dependencies to bundle.properties - additional.bundles. This way we can remove them from the MANIFEST.MF and they will still show up in Eclipse. And add the rest to pom.xml as dependencies. This way maven will see them.

**Comment by cdealti** [comment]
Remember that we want to run these fragments on the target and we don't have Maven there.

**Comment by markoer** [comment]
Sorry, I thought we were only talking about unit tests, here, and that integration tests will remain in their own separate bundles/fragments.

**Comment by cdealti** [comment]
do we really need separate test fragments for IT?

**Comment by markoer** [comment]
The correct Djuro is @djurod ...

**Comment by djurod** [comment]
@cdealti Yes, the other Djuro notified me that he is apparently getting Github notifications meant for me :)

**Comment by djurod** [comment]
Could we resume this discussion?

So just to give a little overview of how we came to have (pure) unit tests in .test bundle fragments. Initially we wanted to have unit tests inside the bundle that we are testing, but the issue that moved us into the direction to have unit tests in the .test bundle fragments was a concern that dependencies for the tests would then also become the dependencies of the bundle (required to use the bundle on the device in the MANIFEST.MF file).

As @markoer suggested in his comments here we can see that our original issue with dependencies can be resolved so we could have pure unit tests directly in the bundle under test.

@cdealti just to be clear, by pure unit tests I mean testing the classes in isolation without dependencies to the platform where the bundle should run. I would consider any test that is supposed to be run on the target platform to be an integration test, so it would belong in the .test bundle fragment...

**Comment by ctron** [comment]
So where is the benefit in that? I only see disadvantages. You would have two different compiler systems running in one project (which already causes troubles), you will have two different testing systems (which probably also causes some issues), you have problems with dependencies (as build.properties is not evaluated by Tycho).

So if we are using Tycho as build environment, an we could definitely change to maven-bundle-plugin or bndtools, but as long as we are using Tycho, we should also use the tools Tycho provide: Using test fragments and running only tycho-surefire.

**Comment by cdealti** [comment]
@ctron do you plan to remove ``src/test/java`` from ``org.eclipse.kura.camel``?

**Comment by ctron** [comment]
I would go for that, yes.

**Comment by cdealti** [comment]
those ``org.eclipse.kura.camel`` tests are Unit Tests and may require access to the internal packages of ``org.eclipse.kura.camel``. Moving them to an ``eclipse-test-plugin`` would require the Fragment-Host header in the test plugin.

**Comment by ctron** [comment]
Yes, it would. That is how test fragments work.

**Comment by ctron** [comment]
If you have a look [1], most of the tests for this bundle already follow this pattern. It is only a single test left which needs to be moved over.

[1] https://github.com/eclipse/kura/tree/develop/kura/org.eclipse.kura.camel.test

**Comment by markoer** [comment]
Let's face it. So long as Kura uses native libraries, there's no way it's only using one build environment, let alone one compile/test system. Only having Tycho brings more problems than it solves and I personally would rather avoid it altogether, but...
Anyway, Tycho means having even unit tests depend on another bloated environment configuration which is rather unstable and behaves differently when run from Maven and when run from an IDE i.e. Eclipse, if it even runs. The whole point of unit tests is that they are bundled with the implementation code (i.e. src/test/java), in the back of developer's mind, and that they can be run quickly without an error-prone setup that Tycho brings. I don't want to run an OSGi container to run a unit test of my code if I don't absolutely have to! The ones implementing the code must want to run the tests, not that they avoid it because "the setup is always broken" or "it's a different project anyway".
Don't mistake this for not wanting to still have some tests run in an approximation of the real environment. But that's what Tycho's test fragments are for and they remain for integration tests. And those shouldn't have to focus on testing every bit of code, but rather on a bit higher level - bundle configuration (i.e. imports, exports, etc.), communication with environment and other services, API functionality, etc.
Conclusion... keep both, but each should serve its own purpose.

**Comment by ctron** [comment]
Of course if you dislike the whole setup of Krua that much, then it hard to follow the way Tycho works. But always working against a well established toolchain will not solved problems but cause many more. There is way Tycho works and as we have seen in the past, working against this creates problems.

So on the long run I would suggest to follow the ideas of Tycho or switch the whole different toolchain. Not just the bits you personally dislike. There are many alternatives. But again, this is a decision which has to be mode on a project level.

**Comment by cdealti** [comment]
@markoer do you need to move something [here](https://github.com/eclipse/kura/wiki/Development-Environment-Tips-and-Tricks) before I close this?

**Comment by markoer** [comment]
The discussion on this topic is still pending in the community meeting. Maybe we should do that first.

**Comment by cdealti** [comment]
@markoer I don't see a lot of discussion about this topic. Closing it. Feel free to re-open it.
